autogpt_platform/backend/backend/data/graph.py

```diff
@@ -312,12 +312,49 @@ def reassign_ids(self, user_id: str, reassign_graph_id: bool = False):
 
     def validate_graph(self, for_run: bool = False):
         def sanitize(name):
-            return name.split("_#_")[0].split("_@_")[0].split("_$_")[0]
+            sanitized_name = name.split("_#_")[0].split("_@_")[0].split("_$_")[0]
+            if sanitized_name.startswith("tools_"):
+                return sanitized_name.split("_")[0]
+            return sanitized_name
+
+        # Validate smart decision maker nodes
+        smart_decision_maker_nodes = set()
+        smd_unique_tool_links = {}
+        agent_nodes = set()
+
+        for node in self.nodes:
+            # Smart decision maker nodes
+            if node.block_id == "3b191d9f-356f-482d-8238-ba04b6d18381":
```

Comment:
We intentionally match stuff based on the block type like below to avoid this hardcoded ID check to happen:
https://github.com/Significant-Gravitas/AutoGPT/pull/9490/files#diff-fe699074de9055342cbf5c2aef8eba0ae08ef41429c2f012789c20eb9dec532cR381

We can get the block object early here.

Reply:
When I did that I was hitting circular import issues. I've hit this problem several times during adding this block...

Reply:
I tweaked this a bit, feel free to check and update

Reply:
This bit of code was deleted